pages/blog/node-js-and-ipv6.mdx

```diff
@@ -0,0 +1,108 @@
+---
+title: "Binding a Node.js Web app to IPv4 and IPv6 Sockets"
+date: "2021-08-05"
+author: "Evan Tahler"
+pullQuote: "How does Grouparoo use Next.js to load pages and components from plugins to modify our web user interface?"
+image: "node-js-and-ipv6/sylvain-gllm-X4dBqRUzO2U-unsplash.jpg"
+tags: [engineering]
+---
+
+We want to make Grouparoo as easy as possible to run, which means considering many different server environments. We recently had a customer who wanted to run Grouparoo in a Docker cluster that only had IPv6 addresses enabled. There are lots of reasons why IPv6 might be better (including the fact that we are [running out of public IPv4 Addresses](https://en.wikipedia.org/wiki/IPv4_address_exhaustion)), but it’s rare to find a deployment environment that _only_ has IPv6 addresses by default. That said, it’s easy to tell your Node.js application to listen to all hosts on both IPv4 and IPv6 - and that's what Grouparoo does [now](https://github.com/grouparoo/grouparoo/pull/2127)!
+
+<Image
+  centered
+  source="https://unsplash.com/photos/X4dBqRUzO2U"
+  alt="Twisty Roads"
+  src="node-js-and-ipv6/sylvain-gllm-X4dBqRUzO2U-unsplash.jpg"
+  width={800}
+  height={1000}
+/>
+
+## The Node.js HTTP Server
+
+When starting a Node.js server, you can to choose both the port to listen on and the hostname to bind to ([docs](https://nodejs.org/api/net.html#net_class_net_server)). Depending on how your server is configured, choosing a specific hostname might route traffic in different ways. Maybe you have 2 network cards (one for internal traffic and one for external), or perhaps you have different networks for IPv4 and IPv6 traffic - choosing a certain hostname may have different effects.
+
+The node HTTP example looks like this:
+
+```ts
+// from https://nodejs.org/en/docs/guides/getting-started-guide/
+const http = require("http");
+
+const hostname = "127.0.0.1";
+const port = 3000;
+
+const server = http.createServer((req, res) => {
+  res.statusCode = 200;
+  res.setHeader("Content-Type", "text/plain");
+  res.end("Hello World");
+});
+
+server.listen(port, hostname, () => {
+  console.log(`Server running at http://${hostname}:${port}/`);
+});
+```
+
+In this example, we are binding only to `127.0.0.1` which is the IPv4 version of what is called a `loopback` - this means that only the same computer can talk to itself. This is a very safe way to test and develop, and a very bad way to run a web server 🤣.
+
+Conversely, what if we wanted to allow traffic in from the widest variety of hosts? This is the default configuration of Grouparoo - we want the application to be as widely available as possible, and the infrastructure to be be in charge of restricting who the application can talk to as needed. In IPv4, that would mean choosing a host of `0.0.0.0` which would allow traffic from anywhere. What about IPv6? It turns out that the string `::` (yes, that’s to colons) means "all" in IPv6, and is actually shorthand for `0.0.0.0.0.0.0.0`.
```

Comment:
```suggestion
Conversely, what if we wanted to allow traffic in from the widest variety of hosts? This is the default configuration of Grouparoo - we want the application to be as widely available as possible, and the infrastructure to be be in charge of restricting who the application can talk to as needed. In IPv4, that would mean choosing a host of `0.0.0.0` which would allow traffic from anywhere. What about IPv6? It turns out that the string `::` (yes, that’s two colons) means "all" in IPv6, and is actually shorthand for `0.0.0.0.0.0.0.0`.
```